src/main/java/com/ishland/c2me/common/optimization/chunkserialization/NbtWriterVisitor.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    public void visitByteArray(byte[] name, NbtByteArray element) {
+        writer.putByteArray(name, element.getByteArray());
+    }
+
+    public void visitByteArray(String name, NbtByteArray element) {
+        this.visitByteArray(NbtWriter.getStringBytes(name), element);
+    }
+
+    public void visitIntArray(byte[] name, NbtIntArray element) {
+        writer.putIntArray(name, element.getIntArray());
+    }
+
+    public void visitIntArray(String name, NbtIntArray element) {
+        this.visitIntArray(NbtWriter.getStringBytes(name), element);
+    }
+
+    public void visitLongArray(byte[] name, NbtLongArray element) {
+        writer.putLongArray(name, element.getLongArray());
+    }
+
+    public void visitLongArray(String name, NbtLongArray element) {
+        this.visitLongArray(NbtWriter.getStringBytes(name), element);
+    }
+
+    public void visitList(byte[] name, NbtList element) {
+        writer.startFixedList(name, element.size(), element.getHeldType());
+        for (NbtElement elementBase : element) {
+            elementBase.accept(this);
+        }
+    }
+
+    public void visitList(String name, NbtList element) {
+        this.visitList(NbtWriter.getStringBytes(name), element);
+    }
+
+    public void visitCompound(byte[] name, NbtCompound compound) {
+        writer.startCompound(name);
+        for (String nameBase : compound.getKeys()) {
+            var element = compound.get(nameBase);
+            visit(nameBase, element);
+        }
+        writer.finishCompound();
+    }
+
+    public void visitCompound(String name, NbtCompound compound) {
+        this.visitCompound(NbtWriter.getStringBytes(name), compound);
+    }
+
+    public void visit(String nameBase, NbtElement element) {
+        switch (element.getType()) {
+            case NbtElement.STRING_TYPE -> visitString(nameBase, (NbtString) element);
+            case NbtElement.BYTE_TYPE -> visitByte(nameBase, (NbtByte) element);
+            case NbtElement.SHORT_TYPE -> visitShort(nameBase, (NbtShort) element);
+            case NbtElement.INT_TYPE -> visitInt(nameBase, (NbtInt) element);
+            case NbtElement.LONG_TYPE -> visitLong(nameBase, (NbtLong) element);
+            case NbtElement.FLOAT_TYPE -> visitFloat(nameBase, (NbtFloat) element);
+            case NbtElement.DOUBLE_TYPE -> visitDouble(nameBase, (NbtDouble) element);
+            case NbtElement.BYTE_ARRAY_TYPE -> visitByteArray(nameBase, (NbtByteArray) element);
+            case NbtElement.INT_ARRAY_TYPE -> visitIntArray(nameBase, (NbtIntArray) element);
+            case NbtElement.LONG_ARRAY_TYPE -> visitLongArray(nameBase, (NbtLongArray) element);
+            case NbtElement.LIST_TYPE -> visitList(nameBase, (NbtList) element);
+            case NbtElement.COMPOUND_TYPE -> visitCompound(nameBase, (NbtCompound) element);
+            default -> throw new IllegalArgumentException("Unknown NbtElement type: " + element.getType());
```

Comment:
is it possible a can a add a new NbtElement type? if so should we fallback to vanilla code?